From 8361ae71adce4cf9408645167734b3e2dbe889f3 Mon Sep 17 00:00:00 2001
From: Veerendranath Jakkam <quic_vjakkam@quicinc.com>
Date: Thu, 4 Jul 2024 15:06:34 +0530
Subject: [PATCH] wpa_supplicant: Add QCA vendor feature flags to indicate RSN
override elements
support Add a separate feature flag for STA mode to indicate support for
RSN override elements.

Patch-dependency: None

Signed-off-by: Veerendranath Jakkam <quic_vjakkam@quicinc.com>
Signed-off-by: Arunpandi Kannan <quic_arunpand@quicinc.com>
---
 src/common/qca-vendor.h | 30 ++++++++++++++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/src/common/qca-vendor.h b/src/common/qca-vendor.h
index 3170cff..a714732 100644
--- a/src/common/qca-vendor.h
+++ b/src/common/qca-vendor.h
@@ -1924,6 +1924,35 @@ enum qca_wlan_vendor_acs_hw_mode {
  * @QCA_WLAN_VENDOR_FEATURE_ENHANCED_AUDIO_EXPERIENCE_OVER_WLAN: Flag indicates
  *	that the device supports enhanced audio experience over WLAN feature.
  * @NUM_QCA_WLAN_VENDOR_FEATURES: Number of assigned feature bits
+ *
+ * @QCA_WLAN_VENDOR_FEATURE_RSN_OVERRIDE_STA: Flag indicates that the device
+ *     supports RSNE/RSNXE overriding in STA mode.
+ *
+ *     For SME offload to the driver case:
+ *     - Supplicant should enable RSNO element use only when the driver
+ *       indicates this feature flag.
+ *     - The driver should enable RSNO element use with the supplicant selected
+ *       BSS only when the supplicant sends an RSNO element with an empty
+ *       payload in the connect request elements buffer in NL80211_CMD_CONNECT.
+ *
+ *     For BSS selection offload to the driver case:
+ *     - Supplicant should enable RSNO element use only when the driver
+ *       indicates this feature flag.
+ *     - Supplicant should always send RSNO elements in the connect request
+ *       elements buffer in NL80211_CMD_CONNECT irrespective of whether RSNO
+ *       elements are supported by the BSS that the supplicant selected
+ *     - The driver should enable RSNO element use only when the supplicant
+ *       sends an RSNO element with an empty payload in connect request
+ *       elements in NL80211_CMD_CONNECT.
+ *     - The driver should remove RSNO elements from the connect request
+ *       elements while preparing the (Re)Association Request frame elements
+ *       if the driver selects a different BSS which is not advertising RSNO
+ *       elements.
+ *
+ *     If both SME and BSS selection offload to the driver, BSS selection
+ *     offload to the driver case rules shall be applied.
+ *
+
  */
 enum qca_wlan_vendor_features {
 	QCA_WLAN_VENDOR_FEATURE_KEY_MGMT_OFFLOAD	= 0,
@@ -1950,6 +1979,7 @@ enum qca_wlan_vendor_features {
 	QCA_WLAN_VENDOR_FEATURE_PROT_RANGE_NEGO_AND_MEASURE_AP = 21,
 	QCA_WLAN_VENDOR_FEATURE_AP_ALLOWED_FREQ_LIST = 22,
 	QCA_WLAN_VENDOR_FEATURE_ENHANCED_AUDIO_EXPERIENCE_OVER_WLAN = 23,
+	QCA_WLAN_VENDOR_FEATURE_RSN_OVERRIDE_STA = 25,
 	NUM_QCA_WLAN_VENDOR_FEATURES /* keep last */
 };
 
-- 
2.34.1

